手动指定内核版本编译

您好，我想请教一下跨内核版本编译的场景，具体而言是在 kernel 4.15.0 中编译 kernel 4.9.0 的 nettrace 工具，目前在编译过程中遇到了以下问题：
```
~/workspace/nettrace/nettrace$ docker run -it --rm --network=host --privileged -v $(pwd):$(pwd) -v /lib/modules/:/lib/modules/ -v /usr/src/:/usr/src/ imagedong/nettrace-build make -C $(pwd) NO_BTF=1 NO_GLOBAL_DATA=1 KERNEL=/home/ubuntu/workspace/nettrace/nettrace/linux all
make: Entering directory '/home/ubuntu/workspace/nettrace/nettrace'
make -C src all
make[1]: Entering directory '/home/ubuntu/workspace/nettrace/nettrace/src'
/bin/sh: 1: pkg-config: not found
python3 gen_trace.py > trace_group.c
python3 gen_trace.py probe > progs/kprobe_trace.h
ln -s vmlinux_header.h kheaders.h
clang -O2 -c -S -Wall -fno-asynchronous-unwind-tables  \
-Wno-incompatible-pointer-types-discards-qualifiers  \
progs/kprobe.c -emit-llvm -Wno-unknown-attributes -I./ -I/home/ubuntu/workspace/nettrace/nettrace/shared/bpf/ -g -DBPF_NO_GLOBAL_DATA -DNO_BTF -D__F_STACK_TRACE -D__F_NFT_NAME_ARRAY -D__KERN_VER=4.9.110 -D__KERN_MAJOR=4 -Wno-unused-function -Wno-compare-distinct-pointer-types -Wuninitialized -D__TARGET_ARCH_x86 -DBPF_NO_PRESERVE_ACCESS_INDEX -nostdinc -isystem /usr/lib/gcc/x86_64-linux-gnu/12/include -I/home/ubuntu/workspace/nettrace/nettrace/linux/arch/x86/include -I/home/ubuntu/workspace/nettrace/nettrace/linux/arch/x86/include/generated -I/home/ubuntu/workspace/nettrace/nettrace/linux/include -I/home/ubuntu/workspace/nettrace/nettrace/linux/arch/x86/include/uapi -I/home/ubuntu/workspace/nettrace/nettrace/linux/arch/x86/include/generated/uapi -I/home/ubuntu/workspace/nettrace/nettrace/linux/include/uapi -I/home/ubuntu/workspace/nettrace/nettrace/linux/include/generated/uapi -include /home/ubuntu/workspace/nettrace/nettrace/linux/include/linux/kconfig.h -I/usr/include/ -D__KERNEL__ -Wno-unused-value -Wno-pointer-sign -Wno-compare-distinct-pointer-types -Wno-gnu-variable-sized-type-not-at-end -Wno-address-of-packed-member -Wno-tautological-compare -Wno-unknown-warning-option -Wno-frame-address -D__F_SK_PRPTOCOL_LEGACY -D__F_NO_SK_FLAGS_OFFSET -Xclang \
-disable-llvm-passes -o - |          \
opt -O2 -mtriple=bpf-pc-linux |       \
llvm-dis |           \
llc -march=bpf -filetype=obj -o progs/kprobe.o
In file included from <built-in>:1:
/home/ubuntu/workspace/nettrace/nettrace/linux/include/linux/kconfig.h:4:10: fatal error: 'generated/autoconf.h' file not found
#include <generated/autoconf.h>
         ^~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make[1]: *** [../common.mk:115: progs/kprobe.o] Error 1
make[1]: Leaving directory '/home/ubuntu/workspace/nettrace/nettrace/src'
make: *** [Makefile:24: all] Error 2
make: Leaving directory '/home/ubuntu/workspace/nettrace/nettrace'
```
作为内核开发的新手，我是直接将 kernel 4.9.0 的源码路径作为 KERNEL 参数，但这明显是不对的，想请教一下如何获得低版本例如 kernel 4.9.0 的头文件，目前 apt-get install 列表已不再维护该内核版本的 linux-headers

而且当我获取了 linux-headers-4.9.0-8-amd64 内核头文件后，指定该路径仍会报错：
```
docker run -it --rm --network=host --privileged -v $(pwd):$(pwd) -v /lib/modules/:/lib/modules/ -v /usr/src/:/usr/src/ imagedong/nettrace-build make -C $(pwd) NO_BTF=1 NO_GLOBAL_DATA=1 KERNEL=/home/ubuntu/workspace/nettrace/nettrace/linux-headers-4.9.0-8-amd64 all
make: Entering directory '/home/ubuntu/workspace/nettrace/nettrace'
make -C src all
make[1]: Entering directory '/home/ubuntu/workspace/nettrace/nettrace/src'
/bin/sh: 1: pkg-config: not found
clang -O2 -c -S -Wall -fno-asynchronous-unwind-tables  \
-Wno-incompatible-pointer-types-discards-qualifiers  \
progs/kprobe.c -emit-llvm -Wno-unknown-attributes -I./ -I/home/ubuntu/workspace/nettrace/nettrace/shared/bpf/ -g -DBPF_NO_GLOBAL_DATA -DNO_BTF -D__F_STACK_TRACE -D__F_NFT_NAME_ARRAY -D__KERN_VER=2.6. -D__KERN_MAJOR=2 -Wno-unused-function -Wno-compare-distinct-pointer-types -Wuninitialized -D__TARGET_ARCH_x86 -DBPF_NO_PRESERVE_ACCESS_INDEX -nostdinc -isystem /usr/lib/gcc/x86_64-linux-gnu/12/include -I/home/ubuntu/workspace/nettrace/nettrace/linux-headers-4.9.0-8-amd64/arch/x86/include -I/home/ubuntu/workspace/nettrace/nettrace/linux-headers-4.9.0-8-amd64/arch/x86/include/generated -I/home/ubuntu/workspace/nettrace/nettrace/linux-headers-4.9.0-8-amd64/include -I/home/ubuntu/workspace/nettrace/nettrace/linux-headers-4.9.0-8-amd64/arch/x86/include/uapi -I/home/ubuntu/workspace/nettrace/nettrace/linux-headers-4.9.0-8-amd64/arch/x86/include/generated/uapi -I/home/ubuntu/workspace/nettrace/nettrace/linux-headers-4.9.0-8-amd64/include/uapi -I/home/ubuntu/workspace/nettrace/nettrace/linux-headers-4.9.0-8-amd64/include/generated/uapi -include /home/ubuntu/workspace/nettrace/nettrace/linux-headers-4.9.0-8-amd64/include/linux/kconfig.h -I/usr/include/ -D__KERNEL__ -Wno-unused-value -Wno-pointer-sign -Wno-compare-distinct-pointer-types -Wno-gnu-variable-sized-type-not-at-end -Wno-address-of-packed-member -Wno-tautological-compare -Wno-unknown-warning-option -Wno-frame-address -D__F_SK_PRPTOCOL_LEGACY -Xclang  \
-disable-llvm-passes -o - |          \
opt -O2 -mtriple=bpf-pc-linux |       \
llvm-dis |           \
llc -march=bpf -filetype=obj -o progs/kprobe.o
<built-in>:1:10: fatal error: '/home/ubuntu/workspace/nettrace/nettrace/linux-headers-4.9.0-8-amd64/include/linux/kconfig.h' file not found
#include "/home/ubuntu/workspace/nettrace/nettrace/linux-headers-4.9.0-8-amd64/include/linux/kconfig.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make[1]: *** [../common.mk:115: progs/kprobe.o] Error 1
make[1]: Leaving directory '/home/ubuntu/workspace/nettrace/nettrace/src'
make: *** [Makefile:24: all] Error 2
make: Leaving directory '/home/ubuntu/workspace/nettrace/nettrace'
```
下面是从 debian 的归档系统中获取的 linux-headers-4.9.0-8-amd64
<img width="1153" alt="image" src="https://github.com/user-attachments/assets/c75b17e6-c147-4a10-ac38-4bf31976890e">

感谢，但在编译过程中还是遇到了新的问题
```
~/workspace/nettrace/nettrace$ docker run -it --rm --network=host --privileged -v $(pwd):$(pwd) -v /lib/modules/:/lib/modules/ -v /usr/src/:/usr/src/ imagedong/nettrace-build make -C $(pwd) NO_BTF=1 NO_GLOBAL_DATA=1 KERNEL=/home/ubuntu/workspace/nettrace/nettrace/linux all
make: Entering directory '/home/ubuntu/workspace/nettrace/nettrace'
make -C src all
make[1]: Entering directory '/home/ubuntu/workspace/nettrace/nettrace/src'
/bin/sh: 1: pkg-config: not found
clang -O2 -c -S -Wall -fno-asynchronous-unwind-tables           \
-Wno-incompatible-pointer-types-discards-qualifiers             \
progs/kprobe.c -emit-llvm -Wno-unknown-attributes -I./ -I/home/ubuntu/workspace/nettrace/nettrace/shared/bpf/ -g -DBPF_NO_GLOBAL_DATA -DNO_BTF -D__F_STACK_TRACE -D__F_NFT_NAME_ARRAY -D__KERN_VER=4.9.110 -D__KERN_MAJOR=4 -Wno-unused-function -Wno-compare-distinct-pointer-types -Wuninitialized -D__TARGET_ARCH_x86 -DBPF_NO_PRESERVE_ACCESS_INDEX -nostdinc -isystem /usr/lib/gcc/x86_64-linux-gnu/12/include -I/home/ubuntu/workspace/nettrace/nettrace/linux/arch/x86/include -I/home/ubuntu/workspace/nettrace/nettrace/linux/arch/x86/include/generated -I/home/ubuntu/workspace/nettrace/nettrace/linux/include -I/home/ubuntu/workspace/nettrace/nettrace/linux/arch/x86/include/uapi -I/home/ubuntu/workspace/nettrace/nettrace/linux/arch/x86/include/generated/uapi -I/home/ubuntu/workspace/nettrace/nettrace/linux/include/uapi -I/home/ubuntu/workspace/nettrace/nettrace/linux/include/generated/uapi -include /home/ubuntu/workspace/nettrace/nettrace/linux/include/linux/kconfig.h -I/usr/include/ -D__KERNEL__ -Wno-unused-value -Wno-pointer-sign -Wno-compare-distinct-pointer-types -Wno-gnu-variable-sized-type-not-at-end -Wno-address-of-packed-member -Wno-tautological-compare -Wno-unknown-warning-option -Wno-frame-address -D__F_SK_PRPTOCOL_LEGACY -D__F_NO_SK_FLAGS_OFFSET -Xclang   \
-disable-llvm-passes -o - |                                     \
opt -O2 -mtriple=bpf-pc-linux |                                 \
llvm-dis |                                                      \
llc -march=bpf -filetype=obj -o progs/kprobe.o
In file included from progs/kprobe.c:94:
progs/core.c:604:17: error: incomplete definition of type 'struct nf_hook_entries'
        num = _(entries->num_hook_entries);
                ~~~~~~~^
/home/ubuntu/workspace/nettrace/nettrace/shared/bpf/skb_parse.h:66:9: note: expanded from macro '_'
        typeof(src) tmp;                                        \
               ^~~
progs/core.c:594:9: note: forward declaration of 'struct nf_hook_entries'
        struct nf_hook_entries *entries;
               ^
progs/core.c:604:17: error: incomplete definition of type 'struct nf_hook_entries'
        num = _(entries->num_hook_entries);
                ~~~~~~~^
/home/ubuntu/workspace/nettrace/nettrace/shared/bpf/skb_parse.h:67:37: note: expanded from macro '_'
        bpf_probe_read_kernel(&tmp, sizeof(src), &(src));       \
                                           ^~~
progs/core.c:594:9: note: forward declaration of 'struct nf_hook_entries'
        struct nf_hook_entries *entries;
               ^
progs/core.c:604:17: error: incomplete definition of type 'struct nf_hook_entries'
        num = _(entries->num_hook_entries);
                ~~~~~~~^
/home/ubuntu/workspace/nettrace/nettrace/shared/bpf/skb_parse.h:67:45: note: expanded from macro '_'
        bpf_probe_read_kernel(&tmp, sizeof(src), &(src));       \
                                                   ^~~
progs/core.c:594:9: note: forward declaration of 'struct nf_hook_entries'
        struct nf_hook_entries *entries;
               ^
progs/core.c:604:6: error: assigning to 'int' from incompatible type 'void'
        num = _(entries->num_hook_entries);
            ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
progs/core.c:610:41: error: incomplete definition of type 'struct nf_hook_entries'
                hooks_event->hooks[i] = (u64)_(entries->hooks[i].hook);
                                               ~~~~~~~^
/home/ubuntu/workspace/nettrace/nettrace/shared/bpf/skb_parse.h:66:9: note: expanded from macro '_'
        typeof(src) tmp;                                        \
               ^~~
progs/core.c:594:9: note: forward declaration of 'struct nf_hook_entries'
        struct nf_hook_entries *entries;
               ^
progs/core.c:610:41: error: incomplete definition of type 'struct nf_hook_entries'
                hooks_event->hooks[i] = (u64)_(entries->hooks[i].hook);
                                               ~~~~~~~^
/home/ubuntu/workspace/nettrace/nettrace/shared/bpf/skb_parse.h:67:37: note: expanded from macro '_'
        bpf_probe_read_kernel(&tmp, sizeof(src), &(src));       \
                                           ^~~
progs/core.c:594:9: note: forward declaration of 'struct nf_hook_entries'
        struct nf_hook_entries *entries;
               ^
progs/core.c:610:41: error: incomplete definition of type 'struct nf_hook_entries'
                hooks_event->hooks[i] = (u64)_(entries->hooks[i].hook);
                                               ~~~~~~~^
/home/ubuntu/workspace/nettrace/nettrace/shared/bpf/skb_parse.h:67:45: note: expanded from macro '_'
        bpf_probe_read_kernel(&tmp, sizeof(src), &(src));       \
                                                   ^~~
progs/core.c:594:9: note: forward declaration of 'struct nf_hook_entries'
        struct nf_hook_entries *entries;
               ^
progs/core.c:610:32: error: operand of type 'void' where arithmetic or pointer type is required
                hooks_event->hooks[i] = (u64)_(entries->hooks[i].hook);
                                             ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/ubuntu/workspace/nettrace/nettrace/shared/bpf/skb_parse.h:64:22: note: expanded from macro '_'
#define _(src)                                                  \
                                                                ^
8 errors generated.
make[1]: *** [../common.mk:115: progs/kprobe.o] Error 1
make[1]: Leaving directory '/home/ubuntu/workspace/nettrace/nettrace/src'
make: *** [Makefile:24: all] Error 2
make: Leaving directory '/home/ubuntu/workspace/nettrace/nettrace'
```

--- a/shared/bpf/skb_macro.h
+++ b/shared/bpf/skb_macro.h
@@ -57,6 +57,10 @@
 #define bpf_core_field_offset(type, field) offsetof(type, field)
 #endif
 
+#ifdef __F_NO_PROBE_READ_STR
+#define bpf_probe_read_str bpf_probe_read
+#endif
+
 #define likely(x)		__builtin_expect(!!(x), 1)
 #define unlikely(x)		__builtin_expect(!!(x), 0)
 

--- a/shared/bpf/skb_shared.h
+++ b/shared/bpf/skb_shared.h
@@ -172,6 +172,7 @@ typedef struct {
 	u32	addr;
 	u32	pkt_len_1;
 	u32	pkt_len_2;
+	u32	pad0;
 	u32	saddr_v6[4];
 	u32	daddr_v6[4];
 	u32	addr_v6[4];

--- a/src/Makefile
+++ b/src/Makefile
@@ -55,6 +55,10 @@ ifeq ($(call str_count,include/net/sock.h,__sk_flags_offset),0)
 	BPF_CFLAGS += -D__F_NO_SK_FLAGS_OFFSET
 endif
 
+ifeq ($(call str_count,include/linux/netfilter.h,nf_hook_entries),0)
+	BPF_CFLAGS += -D__F_NO_NF_HOOK_ENTRIES
+endif
+
 ifeq ($(call struct_field,include/net/netfilter/nf_tables.h,nft_chain,\*name),)
 	CFLAGS += -D__F_NFT_NAME_ARRAY
 endif

--- a/src/Makefile
+++ b/src/Makefile
@@ -59,6 +59,10 @@ ifeq ($(call str_count,include/linux/netfilter.h,nf_hook_entries),0)
 	BPF_CFLAGS += -D__F_NO_NF_HOOK_ENTRIES
 endif
 
+ifeq ($(call str_count,include/uapi/linux/bpf.h,bpf_probe_read_str),0)
+	BPF_CFLAGS += -D__F_NO_PROBE_READ_STR
+endif
+
 ifeq ($(call struct_field,include/net/netfilter/nf_tables.h,nft_chain,\*name),)
 	CFLAGS += -D__F_NFT_NAME_ARRAY
 endif

--- a/src/progs/core.c
+++ b/src/progs/core.c
@@ -604,7 +604,7 @@ DEFINE_KPROBE_SKB(nf_hook_slow, 0, 4)
 		return 0;
 	}
 
-#if __KERN_MAJOR != 3
+#ifndef __F_NO_NF_HOOK_ENTRIES
 	DECLARE_EVENT(nf_hooks_event_t, hooks_event)
 	struct nf_hook_entries *entries;
 	int num, i;

--- a/src/progs/core.c
+++ b/src/progs/core.c
@@ -251,7 +251,7 @@ static inline bool trace_mode_latency(bpf_args_t *args)
  *    0: valid and continue
  *    1: valid and return
  */
-static inline int pre_handle_entry(context_info_t *info)
+static inline int pre_handle_entry(context_info_t *info, u16 func)
 {
 	bpf_args_t *args = (void *)info->args;
 	int ret = 0;
@@ -262,7 +262,7 @@ static inline int pre_handle_entry(context_info_t *info)
 	if (args->max_event && args->event_count >= args->max_event)
 		return -1;
 
-	info->func_status = get_func_status(info->args, info->func);
+	info->func_status = get_func_status(info->args, func);
 	if (mode_has_context(args)) {
 		match_val_t *match_val = bpf_map_lookup_elem(&m_matched,
 							     &info->skb);
@@ -290,9 +290,9 @@ static inline int pre_handle_entry(context_info_t *info)
 
 	if (args->func_stats) {
 		if (ret) {
-			update_stats_key(info->func);
+			update_stats_key(func);
 		} else if (!args->has_filter) {
-			update_stats_key(info->func);
+			update_stats_key(func);
 			args->event_count++;
 			ret = 1;
 		} else {

--- a/src/progs/kprobe.c
+++ b/src/progs/kprobe.c
@@ -39,7 +39,8 @@
 			.args = (void *)CONFIG(),		\
 			info_init				\
 		};						\
-		if (pre_handle_entry(&info)) return 0;		\
+		if (pre_handle_entry(&info, INDEX_##name))	\
+			return 0;				\
 		handle_entry_finish(&info,			\
 				    fake__##name(&info));	\
 		return 0;					\
@@ -71,7 +72,8 @@
 			.args = (void *)CONFIG(),		\
 			info_init				\
 		};						\
-		if (pre_handle_entry(&info)) return 0;		\
+		if (pre_handle_entry(&info, INDEX_##name))	\
+			return 0;				\
 		handle_entry_finish(&info,			\
 				    fake__##name(&info));	\
 		return 0;					\

--- a/src/progs/tracing.c
+++ b/src/progs/tracing.c
@@ -37,7 +37,8 @@
 			.retval = __retval,			\
 			info_init				\
 		};						\
-		if (pre_handle_entry(&info)) return 0;		\
+		if (pre_handle_entry(&info, INDEX_##name))	\
+			return 0;				\
 		handle_entry_finish(&info, fake__##name(&info));\
 		return 0;					\
 	}							\
@@ -50,7 +51,8 @@
 			.args = (void *)CONFIG(),		\
 			info_init				\
 		};						\
-		if (pre_handle_entry(&info)) return 0;		\
+		if (pre_handle_entry(&info, INDEX_##name))	\
+			return 0;				\
 		handle_entry_finish(&info, fake__##name(&info));\
 		return 0;					\
 	}							\
@@ -88,7 +90,8 @@
 			.args = (void *)CONFIG(),		\
 			info_init				\
 		};						\
-		if (pre_handle_entry(&info)) return 0;		\
+		if (pre_handle_entry(&info, INDEX_##name))	\
+			return 0;				\
 		handle_entry_finish(&info, fake__##name(&info));\
 		return 0;					\
 	}							\

